src/sentry/replays/post_process.py

```diff
@@ -2,23 +2,87 @@
 
 import collections
 from itertools import zip_longest
-from typing import Any, Generator, Iterable, Iterator, MutableMapping
-
-
-def process_raw_response(response: list[dict[str, Any]], fields: list[str]) -> list[dict[str, Any]]:
+from typing import Any, Dict, Generator, Iterable, Iterator, List, MutableMapping, Optional, Union
+
+from typing_extensions import TypedDict
+
+
+class DeviceResponseType(TypedDict, total=False):
+    name: Optional[str]
+    brand: Optional[str]
+    model: Optional[str]
+    family: Optional[str]
+
+
+class SDKResponseType(TypedDict, total=False):
+    name: Optional[str]
+    version: Optional[str]
+
+
+class OSResponseType(TypedDict, total=False):
+    name: Optional[str]
+    version: Optional[str]
+
+
+class BrowserResponseType(TypedDict, total=False):
+    name: Optional[str]
+    version: Optional[str]
+
+
+class UserResponseType(TypedDict, total=False):
+    id: Optional[str]
+    username: Optional[str]
+    email: Optional[str]
+    ip: Optional[str]
+    display_name: Optional[str]
+
+
+class ReplayDetailsResponse(TypedDict, total=False):
+    id: str
+    project_id: str
+    trace_ids: List[str]
+    error_ids: List[str]
+    environment: Optional[str]
+    tags: Union[Dict[str, List[str]], List]
+    user: UserResponseType
+    sdk: SDKResponseType
+    os: OSResponseType
+    browser: BrowserResponseType
+    device: DeviceResponseType
+    is_archived: Optional[bool]
+    urls: Optional[List[str]]
+    clicks: List[Dict[str, Any]]
+    count_dead_clicks: Optional[int]
+    count_rage_clicks: Optional[int]
+    count_errors: Optional[int]
+    duration: Optional[int]
+    finished_at: Optional[str]
+    started_at: Optional[str]
+    activity: Optional[int]
+    count_urls: Optional[int]
+    replay_type: str
+    count_segments: Optional[int]
+    platform: Optional[str]
+    releases: List[str]
+    dist: Optional[str]
+
+
+def process_raw_response(
+    response: List[Dict[str, Any]], fields: List[str]
+) -> List[ReplayDetailsResponse]:
     """Process the response further into the expected output."""
     return list(generate_restricted_fieldset(fields, generate_normalized_output(response)))
 
 
 def generate_restricted_fieldset(
-    fields: list[str] | None,
-    response: Generator[dict[str, Any], None, None],
-) -> Iterator[dict[str, Any]]:
+    fields: Optional[List[str]],
+    response: Generator[ReplayDetailsResponse, None, None],
+) -> Iterator[ReplayDetailsResponse]:
 
     """Return only the fields requested by the client."""
     if fields:
         for item in response:
-            yield {field: item[field] for field in fields}
+            yield {field: item[field] for field in fields}  # type: ignore
```

Comment:
there may be a way to fix so we don't have to type ignore, but reading here https://github.com/python/mypy/issues/7178 seems like guaranteeing dictionary key access of TypeDicts are annoying. @getsentry/python-typing anyone here have ideas?

Reply:
the `fields` should be a `list[Literal['possible', 'keys', 'here']] | None` I believe

though it feels weird to only return some of the fields like this -- are there explicit subsets that we can type directly rather than allowing _any_ combination of fields to be sliced out?

Reply:
can we just give them everything and remove / deprecate that behaviour?  what's the risk in giving everything?

Reply:
unfortunately some fields result in expensive clickhouse queries, and it's better to leave them off by default.

Reply:
maybe I'm misreading this -- but haven't they already been fetched at this point?

Reply:
right, yeah. we set default values below if none were returned from the query. we may be able to refactor to make a bit more type friendly/clear. thanks